pkg/quarterdeck/tokens/cache.go

```diff
@@ -0,0 +1,159 @@
+package tokens
+
+import (
+	"container/list"
+	"sync"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+)
+
+// Cache is an in-memory cache that stores issued token claims by ULID.
+type Cache struct {
+	sync.RWMutex
+
+	// The doubly linked list allows us to remove expired tokens without iterating over
+	// the entire list.
+	items *list.List
+
+	// A map to quickly find tokens in the cache.
+	index map[string]*list.Element
+
+	// Maximum number of tokens in the cache.
+	maxTokens uint32
+}
+
+// SessionToken is a string token issued to users that has a limited lifetime. The
+// cache is responsible for checking whether tokens are expired and removing expired
+// tokens to maintain the cache size.
+type SessionToken struct {
+	// ID of the user.
+	userID ulid.ULID
+
+	// Project ID of the project the user is accessing.
+	projectID ulid.ULID
+
+	// Expiration time of the token.
+	expiresAt time.Time
+
+	// The actual token string issued to the user.
+	tks string
+}
+
+// The key is used to lookup the token in the cache by user ID and project ID.
+func (t *SessionToken) Key() string {
+	return t.userID.String() + t.projectID.String()
+}
+
+// Create a new token cache with a maximum number of items.
+func NewCache(size uint32) *Cache {
+	return &Cache{
+		items:     list.New(),
+		index:     make(map[string]*list.Element, 0),
+		maxTokens: size,
+	}
+}
+
+// Get the token from the user ID and project ID or return an error if it does not
+// exist.
+func (c *Cache) Get(userID, projectID ulid.ULID) (tks string, err error) {
+	var (
+		element *list.Element
+		ok      bool
+	)
+
+	// Lookup the token from the cache.
+	token := &SessionToken{
+		userID:    userID,
+		projectID: projectID,
+	}
+
+	c.Lock()
+	defer c.Unlock()
+	if element, ok = c.index[token.Key()]; !ok {
+		return "", ErrCacheMiss
+	}
+
+	// Check if the token has expired. This will panic if the element in the cache is
+	// not a SessionToken so external packages should not access the cache directly.
+	token = element.Value.(*SessionToken)
+	if token.expiresAt.Before(time.Now()) {
+		// If expired, remove the token from the cache and all tokens before it. This
+		// works because token expirations are monotonically increasing.
+		for element != nil {
+			prev := element.Prev()
+			c.remove(element)
+			element = prev
+		}
+
+		return "", ErrCacheExpired
```

Comment:
Could we move this to its own function, e.g. `trunc` so we can benchmark it later?

Reply:
Will do!